qmail-rcptcheck.c

```diff
@@ -0,0 +1,167 @@
+#include <unistd.h>
+#include "control.h"
+#include "env.h"
+#include "stralloc.h"
+#include "substdio.h"
+#include "wait.h"
+
+static char outbuf[128];
+static struct substdio ssout = SUBSTDIO_FDBUF(write,1,outbuf,sizeof outbuf);
+static char errbuf[128];
+static struct substdio sserr = SUBSTDIO_FDBUF(write,2,errbuf,sizeof errbuf);
+
+struct rcptcheck_api {
+  char *(*get_sender)();
+  char *(*get_recipient)();
+  void (*accept_recipient)();
+  void (*reject_recipient)();
+  void (*unable_to_verify)();
+  void (*unable_to_execute)();
+} rcptcheck;
+
+static char *rcptcheck_get_sender() {
+  return env_get("SENDER");
+}
+
+static char *rcptcheck_get_recipient() {
+  return env_get("RECIPIENT");
+}
+
+static void rcptcheck_accept_recipient() {
+  _exit(0);
+}
+
+static void rcptcheck_reject_recipient() {
+  _exit(100);
+}
+
+static void rcptcheck_unable_to_verify() {
+  _exit(111);
+}
+
+static void rcptcheck_unable_to_execute() {
+  _exit(120);
+}
+
+static void run_rcptchecks_under_rcptcheck() {
+  rcptcheck.get_sender        = rcptcheck_get_sender;
+  rcptcheck.get_recipient     = rcptcheck_get_recipient;
+  rcptcheck.accept_recipient  = rcptcheck_accept_recipient;
+  rcptcheck.reject_recipient  = rcptcheck_reject_recipient;
+  rcptcheck.unable_to_verify  = rcptcheck_unable_to_verify;
+  rcptcheck.unable_to_execute = rcptcheck_unable_to_execute;
+}
+
+static void putsdie(substdio *ss,char *string) {
+  substdio_puts(ss,string);
+  substdio_flush(ss);
+  _exit(0);
+}
+
+static char *spp_get_sender() {
+  return env_get("SMTPMAILFROM");
+}
+
+static char *spp_get_recipient() {
+  return env_get("SMTPRCPTTO");
+}
+
+static void spp_accept_recipient() {
+  putsdie(&ssout,"");
+}
+
+static void spp_reject_recipient() {
+  putsdie(&ssout,"E553 sorry, no mailbox here by that name. (#5.1.1)\n");
+}
+
+static void spp_unable_to_verify() {
+  putsdie(&ssout,"R421 unable to verify recipient (#4.3.0)\n");
+}
+
+static void spp_unable_to_execute() {
+  putsdie(&ssout,"R421 unable to execute recipient check (#4.3.0)\n");
+}
+
+static void run_rcptchecks_under_spp() {
+  rcptcheck.get_sender        = spp_get_sender;
+  rcptcheck.get_recipient     = spp_get_recipient;
+  rcptcheck.accept_recipient  = spp_accept_recipient;
+  rcptcheck.reject_recipient  = spp_reject_recipient;
+  rcptcheck.unable_to_verify  = spp_unable_to_verify;
+  rcptcheck.unable_to_execute = spp_unable_to_execute;
+}
+
+static void run_rcptcheck(char *program)
+{
+  char *rcptcheck_program[2] = { program, 0 };
+  int pid;
+  int wstat;
+
+  switch(pid = fork()) {
+    case -1:
+      rcptcheck.unable_to_execute();
```

Comment:
The callbacks are expected to be noreturn here, otherwise one would fallthrough. Without annotations this will lead to compiler warnings.

I wonder if it isn't less error prone to expect the callbacks to return and instead exit in the caller. That way an unexpected return from such a callback will not screw up the program logic.